File: backend/src/models/return.model.js
const { Schema, model } = require('mongoose');

// One entry per problem item covered by this return - a single return can
// span multiple items from the same order (Section 6.9).
const returnItemSchema = new Schema(
  {
    orderItemId: { type: Schema.Types.ObjectId, ref: 'OrderItem', required: true },
    productId: { type: Schema.Types.ObjectId, ref: 'Product', required: true },
    quantity: { type: Number, required: true, min: 1 },
    reasonType: {
      type: String,
      enum: ['damaged', 'wrong_item', 'other'],
      required: true,
    },
    // Required only when reasonType = 'other' - enforced in return.service.js.
    customReason: { type: String, default: null },
  },
  { _id: false }
);

const returnSchema = new Schema(
  {
    // unique: only one return per order, ever - enforced at the DB level, not
    // just in the Service (Section 6.9/8).
    orderId: { type: Schema.Types.ObjectId, ref: 'Order', required: true, unique: true },
    pharmacyId: { type: Schema.Types.ObjectId, ref: 'Pharmacy', required: true },
    warehouseId: { type: Schema.Types.ObjectId, ref: 'Warehouse', required: true },
    items: { type: [returnItemSchema], required: true },
    // One overall note and one shared photo set for the whole return, not per item.
    notes: { type: String, default: null },
    images: { type: [String], default: [] },
    // Money-Flow V2: approving a return CREDITS the pharmacy's account.
    //
    // V1 approved by creating a zero-priced replacement order and moving no
    // money at all, so a pharmacy that handed goods back kept the full bill -
    // and if the product could not be reshipped the approval failed outright,
    // leaving reject as the only option. The replacement mechanism is gone;
    // there is exactly one financial outcome now, and reject still carries a note.
    status: { type: String, enum: ['pending', 'approved', 'rejected'], default: 'pending' },
    rejectionNote: { type: String, default: null },
    resolvedBy: { type: Schema.Types.ObjectId, ref: 'User', default: null },
    resolvedAt: { type: Date, default: null },

    // The `return_credit` ledger entry this return produced, set on approval.
    // The authoritative "only one credit per return" guard is the unique
    // partial index on LedgerEntry.source.returnId; this is for cheap reads.
    creditEntryId: { type: Schema.Types.ObjectId, ref: 'LedgerEntry', default: null },
    // The credited amounts, frozen at approval, and the full working behind
    // them (returnCredit.service.js's breakdown) - so the statement, the
    // invoice and any later dispute all read the same numbers rather than
    // recomputing them against whatever the catalog and the rate say later.
    creditSyp: { type: Number, default: null },
    creditUsd: { type: Number, default: null },
    creditValuation: { type: Schema.Types.Mixed, default: null },
  },
  { timestamps: true }
);

// Level 2 (see docs/PERFORMANCE_OPTIMIZATION.md). The warehouse returns queue
// paginates on _id descending (listPaginatedReturnsForWarehouse):
//   with a status filter -> {warehouseId,status,_id:-1}
//   without one          -> {warehouseId,_id:-1}
// The `_id:-1` suffix on the first also supersedes the old {warehouseId,status}
// (its {warehouseId,status} prefix still serves the unpaginated list + the
// "does this order have a return" lookup) - dropped by the migration script.
returnSchema.index({ warehouseId: 1, status: 1, _id: -1 });
returnSchema.index({ warehouseId: 1, _id: -1 });
returnSchema.index({ pharmacyId: 1 });

module.exports = model('Return', returnSchema);

File: phoenix/lib/features/returns/data/models/return_model.dart
// Section 8: reasonType is a short, fixed enum by design - the pharmacist
// never picks a resolution (approve/reject), only the warehouse does, once
// status leaves 'pending' (Section 7).
const List<String> kReturnReasonTypes = ['damaged', 'wrong_item', 'other'];

class ReturnItemModel {
  const ReturnItemModel({
    required this.orderItemId,
    required this.productId,
    required this.quantity,
    required this.reasonType,
    this.customReason,
    this.productNameAr,
    this.productNameEn,
  });

  final String orderItemId;
  final String productId;
  final int quantity;
  final String reasonType;
  final String? customReason;
  final String? productNameAr;
  final String? productNameEn;

  factory ReturnItemModel.fromJson(Map<String, dynamic> json) => ReturnItemModel(
    orderItemId: json['orderItemId'] as String,
    productId: json['productId'] as String,
    quantity: json['quantity'] as int,
    reasonType: json['reasonType'] as String,
    customReason: json['customReason'] as String?,
    productNameAr: json['productNameAr'] as String?,
    productNameEn: json['productNameEn'] as String?,
  );
}

class ReturnModel {
  const ReturnModel({
    required this.id,
    required this.orderId,
    required this.items,
    this.notes,
    this.images = const [],
    required this.status,
    this.rejectionNote,
    this.replacementOrderId,
    required this.createdAt,
    this.orderNumber,
  });

  final String id;
  final String orderId;
  final List<ReturnItemModel> items;
  final String? notes;
  final List<String> images;
  final String status;
  final String? rejectionNote;
  final String? replacementOrderId;
  final DateTime createdAt;

  // Only present on the list response (joined in server-side).
  final int? orderNumber;

  bool get isPending => status == 'pending';
  bool get isApproved => status == 'approved';
  bool get isRejected => status == 'rejected';

  factory ReturnModel.fromJson(Map<String, dynamic> json) => ReturnModel(
    id: json['id'] as String,
    orderId: json['orderId'] as String,
    items: (json['items'] as List)
        .map((e) => ReturnItemModel.fromJson(e as Map<String, dynamic>))
        .toList(),
    notes: json['notes'] as String?,
    images: json['images'] != null ? List<String>.from(json['images'] as List) : const [],
    status: json['status'] as String,
    rejectionNote: json['rejectionNote'] as String?,
    replacementOrderId: json['replacementOrderId'] as String?,
    createdAt: DateTime.parse(json['createdAt'] as String),
    orderNumber: json['orderNumber'] as int?,
  );
}
